### PR TITLE
feat(bigquery): make `to_pyarrow_batches()` smarter

### PR DESCRIPTION
**Pull request by cpcloud**
This PR makes use of bigquery APIs to allow proper record batch emission without first materializing into a pyarrow Table.

### CODE REVIEW

**Comment by cpcloud** [comment]
BigQuery backend looks good:

```
…/ibis on  improve-bigquery-batching is 📦 v5.0.0 via 🐍 v3.10.10 via ❄️  impure (ibis-3.10.10-env) took 24s
❯ pytest -m bigquery -n auto --dist=loadgroup --snapshot-update --allow-snapshot-deletion -q -x
bringing up nodes...
..x..xxx..x....x..xxx.....x....xx.xxxx.x.x.xxxxxxxxxx..x..x.xxxxxxx.x.xx..xx....xx......x.x.xxxxxxxxxx...xx.x..x.......x...x.......................x......x.....xx......x..x.................. [ 16%]
x...........................x.x....x.x.....x..x......x.....x..xx....xxxx..xx...............x..x......x.....................................x.................................................. [ 32%]
..............................x....x.x.........x.x............x..x.........x.............x....x.x....x...................xx.....x.x...x.....x..x.x...x.....x.......x........x.x...xx.....xx... [ 49%]
....x..x.......xx.x..x...x.......x..xx...xxxxx.x..xx.x.x.xx.x........xxx.x..x.x...x....x..x.......x.....x...x.x.....xx...x.x......x....x.....................x...x.....xx..x....x.x..........x [ 65%]
.x....x........x................x..............x..x.x......x.xxx..x....x..........x...x..x......................x............................x.....xx.x.xxx..........x.x.....xx.............x. [ 82%]
...............s..................................................................................................................x........................................................... [ 98%]
................                                                                                                                                                                               [100%]
957 passed, 1 skipped, 198 xfailed in 124.61s (0:02:04)
```

File: ibis/backends/bigquery/__init__.py [hunk 1]
```diff
@@ -37,6 +37,7 @@
 
 if TYPE_CHECKING:
     import pyarrow as pa
+    from google.cloud.bigquery.table import RowIterator
```
Comment (gforsyth) [inline ibis/backends/bigquery/__init__.py:40]: I was just about to ask if we wanted to put this here.

**Review by gforsyth — APPROVED** [review]
:bangbang:

File: ibis/backends/bigquery/__init__.py [hunk 2]
```diff
@@ -395,14 +405,19 @@ def to_pyarrow_batches(
...
 
-        # kind of pointless, but it'll work if there's enough memory
         query_ast = self.compiler.to_ast_ensure_limit(expr, limit, params=params)
         sql = query_ast.compile()
         cursor = self.raw_sql(sql, params=params, **kwargs)
-        table = self._cursor_to_arrow(cursor)
-        return table.to_reader(chunk_size)
+        batch_iter = self._cursor_to_arrow(
+            cursor,
+            method=lambda result: result.to_arrow_iterable(),
```
Comment (jcrist) [inline ibis/backends/bigquery/__init__.py:417]: Does the iterator returned by `to_arrow_iterable` hold references to the cursor/connection/whatever needed to keep things working, or do we need to do the `weakref` trick we did for `duckdb` here too?

Comment (cpcloud) [reply ibis/backends/bigquery/__init__.py:417]: I'm not sure if the model maps onto the same kind of closable cursor that the sqlalchemy backends have.

Comment (cpcloud) [reply ibis/backends/bigquery/__init__.py:417]: @tswast Should we be judicious about cleaning up the cursor here?

Comment (cpcloud) [reply ibis/backends/bigquery/__init__.py:417]: It looks like the bigquery cursor is purely a compatibility layer.

This iterator is an http iterator, whose connection state semantics I don't fully understand.

Comment (cpcloud) [reply ibis/backends/bigquery/__init__.py:417]: It looks like `HTTPIterator` has no connection state, and is making a get request per page of results.

Comment (tswast) [reply ibis/backends/bigquery/__init__.py:417]: As far as cleanup goes, my hope is that when the BQ Storage client goes out-of-scope it should clean itself up, but there might be some open issues there in the BigQuery client.

Comment (cpcloud) [reply ibis/backends/bigquery/__init__.py:417]: Searching around the Storage Client issue tracker I couldn't find any connection issues. Let's leave it as is for now.

**Review by jcrist — APPROVED** [review]
With the clarification about the cursor/conn lifetime above, I think this is good to go now.

**Pull request merged**